### ISSUE TITLE
Label Datasets for Case Prediction Tasks

### ISSUE THREAD

**Issue by judejudo** [openchlai/ai#48]
Description:
Create and label datasets from preprocessed text and transcription data for case classification and prediction tasks.

Acceptance Criteria:
Datasets are accurately labeled and structured for model input.
Documentation on labeling methodology is provided.
Quality checks are performed to ensure dataset consistency.
Datasets are stored in an accessible repository (e.g., S3, GCS, or project DB).

Tasks:

- [ ] Extract relevant features from preprocessed text data.
-
- [ ] Define labeling guidelines for case classification (e.g., violence, abuse).
- 
- [ ] Perform manual/automated labeling with validation checks.
- 
- [ ] Store datasets in a version-controlled format (e.g., CSV, Parquet).

**Comment by openchlai** [comment]
Classifications 

Case Categories = 
Needs Assessment



Entity Recognitions 
Names...
Location 
Age ...

**Comment by Rogendo** [comment]
# Data Labeling & Synthetic Data Generation Guidelines
This document outlines the comprehensive guidelines for data labeling and synthetic data generation for datasets related to social protection case classifications, needs assessment and entity recognition models. It aims to ensure consistency, quality, and ethical considerations throughout the data lifecycle.



## 1. Purpose & Scope
This document governs:
- Case classification taxonomy and labeling procedures.
- Entity type definitions and annotation standards.
- Synthetic data generation protocols, particularly using Large Language Models (LLMs).
- Quality assurance (QA) processes for both labeled and synthetic data.
- Dataset statistics tracking and reporting.
- Ethical considerations for data handling, generation, and use.

## 2. Case Classification System

### 2.1 Primary Categories
These categories define the overall nature of a case narrative. Each narrative should be assigned one primary category.
| Label               | Definition                                      | Examples                                      | Target Distribution | Notes                                                                 |
|---------------------|-------------------------------------------------|-----------------------------------------------|---------------------|-----------------------------------------------------------------------|
| `child_abuse`       | Harm/risk to minors (<18)                       | Physical abuse, neglect, emotional abuse, sexual exploitation | __%                 | Focus on clear indicators of harm or immediate risk.                  |
| `gender_abuse`      | Gender-based violence/discrimination            | Sexual harassment, domestic violence (gendered), forced marriage, FGM | __%                 | Can overlap with `domestic_abuse` but with a clear gendered dimension. |
| `child_labor`       | Illegal/exploitative child employment           | Hazardous factory work, bonded labor, forced begging | __%                 | Distinguish from acceptable child work (e.g., light chores).         |
| `domestic_abuse`    | Violence/abuse within domestic relationships    | Spousal abuse, elder mistreatment, abuse by family members | __%                 | Includes physical, emotional, financial abuse.                        |
| `non_violence`      | Non-violent support/assistance cases            | Custody disputes, requests for counseling, housing aid, legal aid | __%                 | Cases requiring social support without immediate violence indicators. |

### 2.2 Demographic Controls for Synthetic Generation
These parameters guide the generation of synthetic narratives to ensure diverse representation.
```yaml
gender_distribution: # Refers to individuals mentioned in narratives (victims, perpetrators, others)
  female: 70%
  male: 30%
  other/unspecified: 5%

age_groups: # Approximate distribution for key individuals in narratives
  child: 0-17 (40%)
  adult: 18-60 (55%)
  senior: 61+ (5%)

country_names:
  male
  female
geographic_diversity: # Ensure variety in settings
  urban: 60%
  rural: 30%
  peri_urban: 10%

```

## 3. Entity Annotation Standards
### 3.1 Core Entity Types

These entities should be identified and labeled within the text narratives.
| Label | Scope | Examples |
|-----------------------|-----------------------------------------------------------------------|--------------------------------------------------------------------------|
| PERSON | Individuals involved (victims, perpetrators, reporters, family members) | "the mother", "Mr. Mwangi", "the child", "her husband" |
| AGE | Specific ages or age groups mentioned | "5 years old", "teenager", "elderly woman", "around 30" |
| GENDER | Explicit or strongly implied gender of individuals | "boy", "woman", "father" (implies male) |
| LOCATION | Geographical places relevant to the case | "Nairobi", "Kibera slum", "the local school", "at home" |
| ORGANIZATION | Institutions, agencies, groups involved | "police", "children's services", "local NGO", "the hospital" |
| ABUSE_TYPE | Specific types of abuse or harm described | "physical assault", "neglect", "sexual harassment", "forced labor" |
| RELATIONSHIP | Relationship between individuals mentioned | "mother of the child", "husband and wife", "teacher", "neighbor" |
| EVIDENCE_PHRASE | Phrases indicating reported information or observations | "she reported that", "witnesses saw", "medical report confirmed" |
| TIME_REFERENCE | Phrases indicating when events occurred or duration | "last night", "for three years", "recently", "on 15th March" |
| NEEDS_ASSESSMENT | Explicit statements of needs or required support | "requires shelter", "needs counseling", "medical attention needed" |

### 3.2 Annotation Principles

- `Annotate Specifically`: Label the smallest meaningful unit that represents the entity.

- `Be Consistent:` Apply labels uniformly according to definitions.

- `Context is Key`: Use surrounding text to determine the correct entity type and scope.

- `Avoid Over-Annotation`: Don't label irrelevant entities. Focus on those pertinent to social protection.


## 4. Synthetic Data Generation

### 4.1 Controlled Parameters for Generation

These parameters are actively managed during the synthetic data generation process:

- **Case Category**: Target specific case classifications from Section 2.1.

- **Demographics**: Gender, age, and socioeconomic hints for key individuals (as per Section 2.2).

- **Location Types**: Urban, rural, peri-urban settings.

- **Names**: Utilize diverse lists of male, female, and culturally relevant names to mitigate name bias and ensure variety. Names should be placeholders and anonymized in final datasets.

- **Severity Levels (Implicit)**: Prompts can guide the LLM to generate scenarios of varying severity (e.g., "a mild case of neglect," "a severe instance of physical abuse").
- **Chain of Thought**: Using the chain of thought technique to accertain that the case narratives generated by LLM simulate or are closer to real world scenarios.

- **Narrative Length**: Approximate length of the generated story.

### 4.2 Data Generation Workflow

- **Scenario Definition & Prompt Engineering**: Define specific scenarios based on target categories, demographics, and desired entities. Craft detailed prompts for the LLM.

- **LLM Narrative Generation**: Use LLMs (Mistral model) to generate case narratives based on prompts.

- **Automated Pre-processing & Formatting**: Clean LLM output (e.g., remove generation artifacts, standardize format).

- **Auto-Annotation (Optional)**: Use another LLM or a pre-trained NER model to suggest initial entity annotations.

- **Structured Review for Coherence & Plausibility**: Automated and manual checks for logical consistency, absence of hallucinations, and real-world plausibility.

- **Bias & Fairness Audit**: Review generated data for unintended biases (beyond controlled demographics). Chain-of-Thought (CoT) prompting can be used to ask the LLM to self-critique for bias.

- **Human Validation & Refinement**: Human annotators review, correct, and complete annotations, and validate case classifications.

- **Iterative Improvement**: Feedback from validation is used to refine prompts and the generation process.

### 4.3 LLM-Specific Guidelines for Synthetic Data Generation

This section details best practices when using LLMs for generating synthetic case narratives.

#### 4.3.1 Prompt Engineering for Quality & Control

- Specificity is Key: Provide clear, detailed instructions.

- Role-Playing: "You are a social worker documenting a case. Write a case summary about..."

- Context Setting: Include details about the desired setting, individuals involved (with placeholder roles like "victim," "perpetrator"), and the general situation.

- Desired Output Format: Specify if you need a narrative, a list of events, or text with implicit entities. JSON or structured output can be requested for easier parsing.

- Inclusion/Exclusion Criteria: "The story must include mention of [entity type] but avoid [sensitive topic not relevant]."

- Few-Shot Learning: Provide 2-3 high-quality examples of the desired narrative style, content, and entity density within the prompt.

- Chain-of-Thought (CoT) Prompting: For complex scenarios, instruct the LLM to "think step-by-step" or "explain its reasoning" before generating the final narrative. This can improve coherence and adherence to constraints.

- Iterative Prompt Refinement: Start with a basic prompt, generate a few samples, evaluate, and refine the prompt to improve output quality.

- Temperature and Nucleus Sampling (Top-p): *Lower temperature* (e.g., 0.2-0.5) for more factual, less creative, and more consistent output.
*Higher temperature* (e.g., 0.7-1.0) for more diverse and creative outputs, but with a higher risk of incoherence.
*Adjust top_p* to control the vocabulary pool for token generation.

#### 4.3.2 Addressing Hallucinations and Nonsensical Stories
Hallucinations are plausible-sounding but incorrect or nonsensical statements. Nonsensical stories lack logical coherence.

- Detection Strategies:

    - Human Review: Crucial for nuanced understanding. Implement a "sense check" step where reviewers assess plausibility and coherence.

    - Consistency Checks: For longer narratives, check for internal consistency (e.g., an individual's age shouldn't change erratically).


    - LLM Self-Correction: Prompt the LLM to review its own output: "Review the following story for any logical inconsistencies, factual errors (within the fictional context), or parts that don't make sense. Provide a corrected version."

    - Rule-Based Filters: Simple rules can catch obvious errors (e.g., a 5-year-old perpetrator in a complex financial abuse case).

- Mitigation Strategies:

    - Improved Prompts: More constraints, clearer context, and negative prompts ("Do not include X") can reduce unwanted outputs.

    - Retrieval Augmented Generation (RAG): Provide the LLM with relevant (anonymized) real examples or domain-specific documents as context to ground its generations.

    - Lower Temperature/Top-p: As mentioned, this can reduce randomness.

    - Post-Generation Filtering: Discard narratives flagged by detection strategies.

    - Iterative Editing: For near-misses, human editors can correct minor hallucinations or rephrase incoherent parts.

   

#### 4.3.3 Managing Bias

LLMs can inherit biases from their training data.

- Proactive Prompting for Diversity: Explicitly ask for varied perspectives, backgrounds, and scenarios. "Generate a story about child labor affecting a girl in a rural setting," then "Generate one about a boy in an urban factory."

- Demographic Parameterization: Use the controls in Section 2.2.

- Bias Audits: Regularly review generated datasets for stereotypical portrayals, underrepresentation, or overrepresentation of certain groups in specific roles (e.g., victims, perpetrators) beyond the intended demographic controls.

- LLM Self-Critique for Bias: "Review this story for any potential gender, age, or socioeconomic biases. Suggest ways to make it more neutral or balanced."


#### 4.3.4 Iteration and Human-in-the-Loop

- Generate in Batches: Don't aim for the full dataset in one go. Generate smaller batches.

- Review and Annotate: Have humans review, validate, and annotate these batches.

- Feedback Loop: Use insights from the review process (e.g., common errors, types of hallucinations, recurring biases) to refine prompts, generation parameters, or the overall workflow.

- Domain Expert Consultation: Involve social protection experts to validate the realism and relevance of generated scenarios.

## 5. Quality Assurance (QA)
### 5.1 Automated Checks

- Format Validation: Ensure data conforms to expected schemas (e.g., JSON structure, correct fields).

- Statistical Distribution Checks: Monitor adherence to target demographic and category distributions (Section 2.2, 6.1).

### 5.2 Manual Validation & Review


- Pilot Review: Before large-scale generation/labeling, a small diverse sample (e.g., 50-100 items) is thoroughly reviewed by multiple experts to refine guidelines and identify early issues.

- Sampling for Full Review:

    - Random Sampling: A statistically significant percentage (e.g., 10-20%) of the generated/labeled data undergoes full re-annotation or thorough review by senior validators.

    - Targeted Sampling: Focus on data flagged by automated checks, narratives from new/updated prompts, or complex/ambiguous cases (100% review for these initially).

- Inter-Annotator Agreement (IAA): For labeled data (both human-labeled real data and human-validated synthetic data), ensure IAA to ensure consistency among annotators.

- Error Analysis & Categorization: Systematically categorize errors found during manual review (e.g., incorrect label, missed entity, hallucination). This feeds back into process improvement.

- Drift Report: Track changes in key distributions (categories, demographics, entity density) compared to targets and previous baselines to detect unintended shifts.

### 5.3 Exceptional Error Handling

- Flagging Mechanism: Clear process for annotators/reviewers to flag problematic data points.

- Review & Adjudication: Flagged items are reviewed and adjusted.

- Root Cause Analysis: For systemic errors, conduct RCA to fix the underlying issue in the generation or labeling process such as hallucination.

**Comment by Rogendo** [comment]
[dataset_v1.csv](https://github.com/user-attachments/files/20409686/dataset_v1.csv)        This is the first version of the case category classification dataset that will be used for the case category classification model with a 1000 records

**Comment by Rogendo** [comment]
[LABELING_GUIDELINES.md](https://github.com/user-attachments/files/20409806/LABELING_GUIDELINES.md)  This Markdown document documents the guidelines and the procedure we followed to generate our synthetic data and also to label our data we are to use to create the NLP models.

**Comment by shemmiriam** [comment]
This task is confirmed and uploaded.
